File: leetcode-practice-java/src/main/java/io/hank/leetcode/practices/_006_Zigzag_Conversion.java
package io.hank.leetcode.practices;

import java.util.Arrays;

import io.hank.leetcode.complexity.ComplexityType;
import io.hank.leetcode.complexity.SpaceComplexity;
import io.hank.leetcode.complexity.TimeComplexity;

/**
 * The string "PAYPALISHIRING" is written in a zigzag pattern on a given number of rows like this: (you may want to display this pattern in a fixed font for better legibility)
 * <pre>
 * P   A   H   N
 * A P L S I I G
 * Y   I   R
 * </pre>
 * And then read line by line: "PAHNAPLSIIGYIR"
 * <p>
 * Write the code that will take a string and make this conversion given a number of rows:
 * <br>
 * string convert(string s, int numRows);
 * <p>
 * Example 1:
 * <br>
 * Input: s = "PAYPALISHIRING", numRows = 3
 * <br>
 * Output: "PAHNAPLSIIGYIR"
 * <p>
 * Example 2:
 * <br>
 * Input: s = "PAYPALISHIRING", numRows = 4
 * <br>
 * Output: "PINALSIGYAHRPI"
 * <p>
 * Explanation:
 * <pre>
 * P     I    N
 * A   L S  I G
 * Y A   H R
 * P     I
 * </pre>
 * Example 3:
 * <br>
 * Input: s = "A", numRows = 1
 * <br>
 * Output: "A"
 */
public class _006_Zigzag_Conversion extends LeetcodeProblemSolution {

    @TimeComplexity(ComplexityType.O_N)
    @SpaceComplexity(ComplexityType.O_MIN_M_N)
    String convert(String s, int numRows) {
        if (numRows < 2) {
            return s;
        }

        int n = s.length();
        StringBuilder[] sbArray = new StringBuilder[Math.min(numRows, n)];
        for (int i = 0; i < sbArray.length; i++) {
            sbArray[i] = new StringBuilder();
        }

        int i = 0;

        while (i < n) {
            for (int row = 0; row < numRows && i < n; row++) {
                sbArray[row].append(s.charAt(i++));
            }
            for (int row = numRows - 2; row > 0 && i < n; row--) {
                sbArray[row].append(s.charAt(i++));
            }
        }

        StringBuilder result = new StringBuilder();
        Arrays.stream(sbArray).forEach(sb -> {
            result.append(sb.toString());
        });
        return result.toString();
    }
}

File: leetcode-practice-kotlin/src/main/kotlin/io/hank/leetcode/kotlin/practices/_006_Zigzag_Conversion.kt
package io.hank.leetcode.kotlin.practices

/**
 * The string "PAYPALISHIRING" is written in a zigzag pattern on a given number of rows like this: (you may want to display this pattern in a fixed font for better legibility)
 *
 * P   A   H   N
 *
 * A P L S I I G
 *
 * Y   I   R
 * And then read line by line: "PAHNAPLSIIGYIR"
 *
 * Write the code that will take a string and make this conversion given a number of rows:
 *
 * string convert(string s, int numRows);
 *
 * Example 1:
 *
 * Input: s = "PAYPALISHIRING", numRows = 3
 *
 * Output: "PAHNAPLSIIGYIR"
 *
 * Example 2:
 *
 * Input: s = "PAYPALISHIRING", numRows = 4
 *
 * Output: "PINALSIGYAHRPI"
 *
 * Explanation:
 *
 * P     I    N
 *
 * A   L S  I G
 *
 * Y A   H R
 *
 * P     I
 *
 * Example 3:
 *
 * Input: s = "A", numRows = 1
 *
 * Output: "A"
 */
class _006_Zigzag_Conversion : LeetcodeProblemSolution() {
    fun convert(s: String, numRows: Int): String {
        if (numRows < 2) {
            return s
        }

        val strLen = s.length
        val sbArray = Array(minOf(strLen, numRows)) { StringBuilder() }

        var i = 0
        while (i < strLen) {
            for (row in 0 until numRows) {
                takeIf { i < strLen }?.let { sbArray[row].append(s[i++]) }
            }
            for (row in numRows - 2 downTo 1) {
                takeIf { i < strLen }?.let { sbArray[row].append(s[i++]) }
            }
        }

        return StringBuilder().apply {
            sbArray.forEach { this.append(it) }
        }.toString()
    }
}
